// tfrecord/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// We expose all protobuf.js types, in case someone needs them.
//
// However, they're not part of the stable API, so do not take a dependency on
// them lightly.
var protos_1 = require("./gen/protos");
exports.protos = protos_1.tensorflow;
const protos_2 = require("./gen/protos");
exports.Example = protos_2.tensorflow.Example;
const example_builder_1 = require("./example_builder");
exports.ExampleBuilder = example_builder_1.ExampleBuilder;
const reader_1 = require("./reader");
exports.Reader = reader_1.Reader;
const record_reader_1 = require("./record_reader");
exports.RecordReader = record_reader_1.RecordReader;
const record_writer_1 = require("./record_writer");
exports.RecordWriter = record_writer_1.RecordWriter;
const writer_1 = require("./writer");
exports.Writer = writer_1.Writer;
exports.createBuilder = example_builder_1.ExampleBuilder.create;
exports.createReader = reader_1.Reader.create;
exports.createWriter = writer_1.Writer.create;
