// vott-ct/lib/js/ct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Toolbar_1 = require("./CanvasTools/Toolbar/Toolbar");
const RegionsManager_1 = require("./CanvasTools/Region/RegionsManager");
const PointRegion_1 = require("./CanvasTools/Region/Point/PointRegion");
const RectRegion_1 = require("./CanvasTools/Region/Rect/RectRegion");
const AreaSelector_1 = require("./CanvasTools/Selection/AreaSelector");
const CanvasTools_Filter_1 = require("./CanvasTools/CanvasTools.Filter");
const Rect_1 = require("./CanvasTools/Core/Rect");
const Point2D_1 = require("./CanvasTools/Core/Point2D");
const RegionData_1 = require("./CanvasTools/Core/RegionData");
const Tag_1 = require("./CanvasTools/Core/Tag");
const TagsDescriptor_1 = require("./CanvasTools/Core/TagsDescriptor");
const CanvasTools_Editor_1 = require("./CanvasTools/CanvasTools.Editor");
require("snapsvg-cjs");
var CanvasTools;
(function (CanvasTools) {
    CanvasTools.Core = {
        Rect: Rect_1.Rect,
        Point2D: Point2D_1.Point2D,
        RegionData: RegionData_1.RegionData,
        TagsDescriptor: TagsDescriptor_1.TagsDescriptor,
        Tag: Tag_1.Tag
    };
    CanvasTools.Selection = {
        AreaSelector: AreaSelector_1.AreaSelector,
        SelectionMode: AreaSelector_1.SelectionMode
    };
    CanvasTools.Region = {
        RegionsManager: RegionsManager_1.RegionsManager,
        PointRegion: PointRegion_1.PointRegion,
        RectRegion: RectRegion_1.RectRegion
    };
    CanvasTools.Filters = {
        InvertFilter: CanvasTools_Filter_1.InvertFilter,
        GrayscaleFilter: CanvasTools_Filter_1.GrayscaleFilter
    };
    CanvasTools.Editor = CanvasTools_Editor_1.Editor;
    CanvasTools.Toolbar = Toolbar_1.Toolbar;
})(CanvasTools = exports.CanvasTools || (exports.CanvasTools = {}));
require("./../css/canvastools.css");
