// vott-ct/lib/js/CanvasTools/Core/TagsDescriptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Tag_1 = require("./Tag");
class TagsDescriptor {
    static BuildFromJSON(data) {
        let p = Tag_1.Tag.BuildFromJSON(data.primary);
        let s = (data.secondary === undefined) ? [] : data.secondary.map((tag) => Tag_1.Tag.BuildFromJSON(tag));
        return new TagsDescriptor(p, s);
    }
    get all() {
        return this.allTags.map((tag) => tag.copy());
    }
    get primary() {
        return this.primaryTag.copy();
    }
    get secondary() {
        return this.allTags.filter((tag) => {
            return (tag.name !== this.primary.name);
        });
    }
    constructor(primaryTag, secondaryTags = []) {
        this.allTags = new Array(primaryTag, ...secondaryTags);
        this.primaryTag = primaryTag;
    }
    toString() {
        let str = this.primary.name;
        this.secondary.forEach((tag) => {
            str += " " + tag.name;
        });
        return str;
    }
    toJSON() {
        return {
            primary: this.primaryTag.toJSON(),
            secondary: this.secondary.map((tag) => tag.toJSON())
        };
    }
}
exports.TagsDescriptor = TagsDescriptor;
