// vott-ct/lib/js/CanvasTools/Selection/MaskElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ElementPart_1 = require("./ElementPart");
const RectElement_1 = require("./RectElement");
class MaskElement extends ElementPart_1.ElementPart {
    constructor(paper, boundRect, maskOut) {
        super(paper, boundRect);
        this.maskOut = maskOut;
        this.buildUIElements();
        this.resize(boundRect.width, boundRect.height);
        this.hide();
    }
    resize(width, height) {
        super.resize(width, height);
        this.mask.resize(width, height);
        this.maskIn.resize(width, height);
    }
    buildUIElements() {
        this.mask = this.createMask();
        this.maskIn = this.createMaskIn();
        this.maskOut.node.addClass("maskOutStyle");
        const combinedMask = this.paper.g();
        combinedMask.add(this.maskIn.node);
        combinedMask.add(this.maskOut.node);
        this.mask.node.attr({
            mask: combinedMask,
        });
        this.node = this.mask.node;
    }
    createMask() {
        const r = new RectElement_1.RectElement(this.paper, this.boundRect, this.boundRect);
        r.node.addClass("maskStyle");
        return r;
    }
    createMaskIn() {
        const r = new RectElement_1.RectElement(this.paper, this.boundRect, this.boundRect);
        r.node.addClass("maskInStyle");
        return r;
    }
}
exports.MaskElement = MaskElement;
