// vott-ct/lib/js/CanvasTools/Region/Component/TagsComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RegionComponent_1 = require("./RegionComponent");
class TagsComponent extends RegionComponent_1.RegionComponent {
    constructor(paper, paperRect, regionData, tags, styleId, styleSheet, tagsUpdateOptions) {
        super(paper, paperRect, regionData, null);
        this.styleSheet = null;
        this.styleMap = [];
        this.styleLightMap = [];
        this.styleId = styleId;
        this.styleSheet = styleSheet;
        this.tags = tags;
        this.tagsUpdateOptions = tagsUpdateOptions;
        this.node = paper.g();
        this.node.addClass("tagsLayer");
    }
    initStyleMaps(tags) {
    }
    updateTags(tags, options) {
        this.tags = tags;
        this.tagsUpdateOptions = options;
        this.rebuildTagLabels();
        this.clearStyleMaps();
        this.initStyleMaps(tags);
        let showBackground = (options !== undefined) ? options.showRegionBackground : true;
        this.applyStyleMaps(showBackground);
    }
    rebuildTagLabels() {
    }
    clearStyleMaps() {
        while (this.styleSheet.cssRules.length > 0) {
            this.styleSheet.deleteRule(0);
        }
    }
    applyStyleMaps(showRegionBackground = true) {
        if (this.tags && this.tags.primary !== undefined) {
            window.requestAnimationFrame(() => {
                let sm = (showRegionBackground ? this.styleMap : this.styleLightMap);
                for (let i = 0; i < sm.length; i++) {
                    let r = sm[i];
                    this.styleSheet.insertRule(`${r.rule}{${r.style}}`, 0);
                }
            });
        }
    }
}
exports.TagsComponent = TagsComponent;
